### ISSUE TITLE
Preview feature feedback: `nativeTypes`

### ISSUE THREAD

**Issue by nikolasburk** [prisma/prisma#4045]
Please share your feedback about the preview feature `nativeTypes` that was released in [v2.11.0](https://github.com/prisma/prisma/releases/tag/2.11.0) in this issue.

- If you encounter a bug, please [open a bug report](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=bug_report.md&title=) in this repo. 
- If the feature is working well for you, please share this in a comment below or leave a 👍 on this issue.

If you have any questions, don't hesitate to ask them in the [`#prisma2`](https://app.slack.com/client/T0MQBS8JG/CKQTGR6T0) channel in the [Prisma Slack](https://slack.prisma.io).

**Comment by cyrus-za** [comment]
Does this mean we finally got postgres interval type supported by annotating the schema with @db.interval?

**Comment by 2color** [comment]
@cyrus-za 

The interval type is still not supported in Prisma Client. Currently, the native type annotations don’t change the way that Prisma Client works. 

For the interval type to be supported, we’d need to add a representation for it in Prisma Client as we recently did with BigInt and Decimal.

It’d be great if you could [open an issue](https://github.com/prisma/prisma/issues/new/choose) for interval support and add some examples of how you’re using it so that we can better understand your use-case and how to best implement it.

**Comment by nikolasburk** [comment]
There already is an open issue to support the `interval` type here: https://github.com/prisma/prisma-client-js/issues/917

**Comment by naveen-bharathi** [comment]
@nikolasburk When I use prisma migrate with v2.11.0, it results in an error
```zsh
Error: "nativeTypes" preview feature is not supoorted yet. Remove it from your schema to use Prisma Migrate.
```

**Comment by pantharshit00** [comment]
@naveen-bharathi Migrate doesn't support native types yet. Right now it is meant to be used via introspection.

Read more details here: https://www.prisma.io/docs/concepts/components/preview-features/native-types#limitations

**Comment by albertoperdomo** [comment]
@naveen-bharathi Migrate supports native types since 2.15.0. Have you been able to test it?

**Comment by albertoperdomo** [comment]
@cyrus-za Check out `Unsupported`, added in [2.16.0](https://github.com/prisma/prisma/releases/tag/2.16.0). This gives you the option to have interval type columns in the database, although there won't be support for it in the Client.

**Comment by naveen-bharathi** [comment]
> @naveen-bharathi Migrate supports native types since 2.15.0. Have you been able to test it? 

Works very well. 🙂👌🏻

**Comment by albertoperdomo** [comment]
Hello everyone, 

we are looking to stabilize this feature ASAP and make this available in Prisma for everyone. For this, we are asking for your help!

Can you please share your feedback (positive or negative) and let us know if there are any issues that we should look into before this can be wrapped up?

**Comment by livthomas** [comment]
@albertoperdomo I'm trying to create a new record in an SQLite table where one of the fields is of type `Decimal`. If I pass a negative value as `String` to `create` function, I get the following error from Prisma:
```
Argument initialBalance: Got invalid value '-65.7' on prisma.createOneCashBox. Provided String, expected Decimal.
```
It works if I pass it the same value as `Number` or `Decimal`. Positive `String` values also work without any problems.

**Comment by eddeee888** [comment]
Hello @albertoperdomo, when updating a field with the type `String` to `Unsupported("TEXT")` for MySQL, it keeps trying to create a new migration every time I run `prisma migrate dev --preview-feature`. The content of the new migration file is the same with the existing one. 

It normally does not show a prompt asking for a new migration if `Unsupported` is not used

I'm using prisma@2.16.0. Here's how to replicate this issue:

- Create a new table with a field of type `String` ( mapped to VARCHAR (191) )
- Run `prisma migrate dev --preview-feature`
- Update the field type to `Unsupported("TEXT")`
- Run `prisma migrate dev --preview-feature`. This is working as expected as it creates a new migration file
- Run `prisma migrate dev --preview-feature` again. This will create another migration file - I don't think this behaviour is desirable because we would be trying to apply the same structure update.

**Comment by albertoperdomo** [comment]
Hello @eddeee888,

this might be slightly tangential, but why are you using `Unsupported` to map to `text` instead of the specific notation? For `text` on MySQL, I think it would look like this?

`myField String @db.Text`

Then Prisma Migrate should pick it up and create a `text` column out of the box, or change from `varchar` to `text` if that is the initial state.

Nevertheless, I can't think of a reason why Unsupported would trigger additional migrations. Hard to say with more details and the ability to reproduce. Can you please open a GH issue with the corresponding schema, migrations etc. ?

**Comment by tomhoule** [comment]
I think it's a case sensitivity issue (mysql probably returns `text` lowercased), I'll have a look. Does it still happen if you change the type in your schema to `Unsupported("text")`?

**Comment by eddeee888** [comment]
I found the problem: I didn't have `previewFeatures = ["nativeTypes"]` when using `Unsupported`.. Although I expected it to throw an error if I didn't have the flag on.

When I activate the flag now, it does not create a new migration when I update the type of a field. It only creates new migration for a brand new field. 🤔 

Anyways, as @albertoperdomo pointed out, mapping the type works well for my use case. Thank you 🙌

**Comment by tomhoule** [comment]
@eddeee888 I opened an issue for the problem you reported: https://github.com/prisma/prisma/issues/5493

**Comment by albertoperdomo** [comment]
@livthomas I think you are expected to use the Decimal class, like this: `new Prisma.Decimal(-65.7)`.
Still, the fact that you can pass a String for positive numbers but not negative numbers is still weird though.

Can you please open a bug report for this?

**Comment by livthomas** [comment]
@albertoperdomo Here it is: https://github.com/prisma/prisma/issues/5508

**Comment by naveen-bharathi** [comment]
> @livthomas I think you are expected to use the Decimal class, like this: `new Prisma.Decimal(-65.7)`.
> 
> Still, the fact that you can pass a String for positive numbers but not negative numbers is still weird though.
> 
> 
> 
> Can you please open a bug report for this?

@livthomas @albertoperdomo 
I think negative numbers are not supported by type `decimal`

For more, please read [postgresql.org](https://www.postgresql.org/docs/9.1/datatype-numeric.html)

![image](https://user-images.githubusercontent.com/32498535/107123868-25af7080-68c6-11eb-92f8-a0be304c9df7.png)![image](https://user-images.githubusercontent.com/32498535/107123873-2811ca80-68c6-11eb-9b59-f28108ddbe7e.png)

**Comment by livthomas** [comment]
@naveen-bharathi, you're referring to something completely different here. Read it again, `precision` and `scale` are parameters of `NUMERIC` type. So this is about type definition while I was talking about the actual value I wanted to save to the DB. Of course you can store a negative value in `NUMERIC` type. Otherwise, this SQL type would be completely useless in any financial application.

**Comment by Dremora** [comment]
Is there any reason `BigInt` interface is used instead of `bigint` literal type? Wouldn't it be more preferable to deal with a primitive type, like in case of `number` and `string` (as opposed to `Number` and `String`)?

**Comment by meirianto** [comment]
There needs to be better support for unsupported types, such as manually being able to cast or transform it onto a custom field in Prisma. It's unacceptable that there is no possible way to work with unsupported columns through Prisma Client CRUD operations. I'm trying to have a column with the `point` geometric type, and I would love to be able to cast it to two float fields in the Prisma model even if this type is not supported. Alas, there is no ability to define custom fields into a Prisma model similarly to Nexus-prisma-plugin where one can define still define a custom field on a generated GraphQL object.

**Comment by cyrus-za** [comment]
+1 @meirianto 
Even if we manually need to send a callback function to parse it from what SQL returns, we still need a way. 
Right now for our interval type column we do a separate raw query and convert it in sql and then take that value and merge it in with a prisma query that excludes that column. This is annoying, especially when we fetching a list as we need to loop through each item to add it and also need to make sure both queries are sorted. 

Right now we do this

```ts
// prisma.ts 
const prismaInstance = () => {
  if (instance) return instance
  instance = new PrismaClient()
  return instance
}
prismaInstance().$queryRaw(`SET intervalstyle TO iso_8601;`) // Workaround for prisma not supporting PostgreSQL intervals.
```

```ts 
// services/timeEntry.ts

const findAll = async () => {
 const timeEntries = await prisma.timeEntry.findMany({
    orderBy: [{ startedAt: 'asc' }],
  })

  const timeEntryIds = timeEntries.map((a) => a.id)

  const timeEntryDurations: { id: number; duration: string }[] = await prisma.$queryRaw(oneLine`
      SELECT id, duration::text FROM "timeEntries"
      WHERE id IN (${timeEntryIds});
  `)

  const addDuration = (timeEntry: TimeEntry) => ({
    ...timeEntry,
    duration: flow(
      find({ id: timeEntry.id }),
      get('duration'),
      moment.duration
    )(timeEntryDurations),
  })
}
```

What would be nice is if we could do this instead

```ts
const findAll = async () => {
 const timeEntries = await prisma.timeEntry.findMany({
    orderBy: [{ startedAt: 'asc' }],
    customColumn: {
      duration: {
        castTo: 'text' // adds duration::text in postgres query,
        parse: moment.duration // custom function to parse the output from postgres before returning the promise
      }
    }
  })
}
```

**Comment by janpio** [comment]
Hey @meirianto and @cyrus-za - this is valuable feedback but outside of the scope of the current `nativeTypes` preview feature. It would be best if you could create a new issue with your use case so we can properly discuss this with you and possibly later prioritize it. Thanks.

**Comment by cyrus-za** [comment]
Thanks @janpio I created a issue here: https://github.com/prisma/prisma/issues/5683

**Comment by albertoperdomo** [comment]
Native types was released in General Availability in 2.17.0. Release notes: https://github.com/prisma/prisma/releases/tag/2.17.0

Thanks to everyone that provided feedback!
We are still interested in hearing from you how you would improve the native type capabilities in Prisma through the usual channels (GH issue, #product-feedback on Public Slack).
